app/views/programming_groups/index.html.slim

```diff
@@ -0,0 +1,34 @@
+h1 = ProgrammingGroup.model_name.human(count: 2)
+
+- if params[:exercise_id].nil?
+  = render(layout: 'shared/form_filters') do |f|
+    .col-auto
+      = f.label(:exercise_id_eq, t('activerecord.attributes.programming_group.exercise'), class: 'visually-hidden form-label')
+      = f.collection_select(:exercise_id_eq, Exercise.with_programming_groups, :id, :title, class: 'form-control', prompt: t('activerecord.attributes.programming_group.exercise'))
+
+.table-responsive
+  table.table.mt-4 class="#{@programming_groups.present? ? 'sortable' : ''}"
+    thead
+      tr
+        th = sort_link(@search, :id, t('activerecord.attributes.programming_group.programming_group_name'))
+        th = sort_link(@search, :exercise_id, t('activerecord.attributes.programming_group.exercise'))
+        th = sort_link(@search, :user_id, t('activerecord.attributes.programming_group.member'))
```

Comment:
I tried to add a filter for the user ID with adding the following to code snippets to the index.html.slim:
```
   .col-auto
      = f.label(:external_user_id_cont, t('activerecord.attributes.programming_group.member'), class: 'visually-hidden form-label')
      = f.search_field(:external_user_id_cont, class: 'form-control', placeholder: t('activerecord.attributes.programming_group.member'))
```

and
```
th.sorttable_nosort = sort_link(@search, :external_user_id, t('activerecord.attributes.programming_group.member'))
```

And adding `external_user` to the ransackable_associations. But I get the following error:
```
undefined method `external_user_id_cont' for Ransack::Search<class: ProgrammingGroup, base: Grouping <combinator: and>>:Ransack::Search
```

Reply:
Alright, I see. Indeed, your changes look mostly fine, there is just a minor thing left with the configuration of Ransack (the permitted attributes / associations) that is preventing the search to work correctly.

If I am not mistaken, the following column is not required, is it? In that case (and because it would still have the same error as I mentioned before), I would just omit it:

```slim
th.sorttable_nosort = sort_link(@search, :external_user_id, t('activerecord.attributes.programming_group.member'))
```

Besides that, let's dig into further details of the `external_user_id_cont`:

1. The `_cont` ending is not ideal. Since we are dealing with an ID column, what does `cont` mean here semantically? Is `4` contained in the ID `42`? As usual, I would switch to `eq` for ID columns.
2. You are trying to access the `external_user` information here. This leaves out all internal users. Further, no direct access is available (but rather only through the programming group memberships). Hence, we always need to prefix with `programming_group_memberships_`. There are two solutions I can think of, which are outlined in the following.

- Either, allow searching for any user ID. Entering the member ID 42 would hence list all programming groups where either internal user 42 or external user 42 is a member of. In that case, the variable should be `programming_group_memberships_user_id_eq`.
- Alternatively, provide two distinct search fields, one for internal and one for external users. This allows admins to search specifically. Then, we need two search fields, one named `programming_group_memberships_user_of_ExternalUser_type_id_eq` and the other `programming_group_memberships_user_of_InternalUser_type_id_eq`. This is called a polymorphic search and further outlined [in the docs](https://activerecord-hackery.github.io/ransack/going-further/polymorphic-search/). In this second case, we also need to allow searching for the user in the `model/programming_group_membership.rb` with
  ```rb
  def self.ransackable_associations(_auth_object = nil)
    %w[user]
  end
  ```
  and further add the `id` to `ransackable_attributes` in `model/user.rb`.

With any of these two methods, I was able to search successfully for corresponding users.

Reply:
Awesome, thank you! Works perfectly fine 💪